app/gui2/src/components/GraphEditor/collapsing.ts

```diff
@@ -0,0 +1,266 @@
+import { GraphDb } from '@/stores/graph/graphDatabase'
+import { Ast } from '@/util/ast'
+import { unwrap } from '@/util/data/result'
+import { tryIdentifier, type Identifier } from '@/util/qualifiedName'
+import assert from 'assert'
+import * as set from 'lib0/set'
+import { IdMap, type ExprId } from 'shared/yjsModel'
+
+// === Types ===
+
+/** Information about code transformations needed to collapse the nodes. */
+interface CollapsedInfo {
+  extracted: ExtractedInfo
+  refactored: RefactoredInfo
+}
+
+/** The information about the extracted function. */
+interface ExtractedInfo {
+  /** Nodes with these ids should be moved to the function body, in their original order. */
+  ids: Set<ExprId>
+  /** The output information of the function. */
+  output: Output | null
+  /** The list of extracted function’s argument names. */
+  inputs: Identifier[]
+}
+
+/** The information about the output value of the extracted function. */
+interface Output {
+  /** The id of the node the expression of which should be replaced by the function call.
+   * This node is also included into `ids` of the {@link ExtractedInfo} and must be moved into the extracted function.
+   */
+  node: ExprId
+  /** The identifier of the return value of the extracted function. */
+  identifier: Identifier
+}
+
+/** The information about the refactored node, the one that needs to be replaced with the function call. */
+interface RefactoredInfo {
+  /** The id of the refactored node. */
+  id: ExprId
+  /** The pattern of the refactored node. Included for convinience, collapsing does not affect it. */
+  pattern: string
+  /** The new expression of the refactored node. A call to the extracted function with the list of necessary arguments. */
+  expression: string
+}
+
+// === prepareCollapsedInfo ===
+
+/** Prepare the information necessary for collapsing nodes.
+ * @throws errors in case of failures, but it should not happen in normal execution.
+ */
+export function prepareCollapsedInfo(selected: Set<ExprId>, graphDb: GraphDb): CollapsedInfo {
+  if (selected.size == 0) throw new Error('Collapsing requires at least a single selected node.')
+  // Leaves are the nodes that have no outgoing connection.
+  const leaves = new Set([...selected])
+  const inputs: Identifier[] = []
+  let output: Output | null = null
+  for (const [targetExprId, sourceExprIds] of graphDb.connections.allReverse()) {
+    const target = graphDb.getExpressionNodeId(targetExprId)
+    if (target == null) throw new Error(`Connection target node for id ${targetExprId} not found.`)
+    for (const sourceExprId of sourceExprIds) {
+      const source = graphDb.getPatternExpressionNodeId(sourceExprId)
+      if (source == null)
+        throw new Error(`Connection source node for id ${sourceExprId} not found.`)
+      const startsInside = selected.has(source)
+      const endsInside = selected.has(target)
+      const stringIdentifier = graphDb.nodeIdToNode.get(source)?.pattern?.repr()
```

Comment:
This is not an identifier, but an entire pattern. `getOutputPortIdentifier` should be used instead.